Atomic/push.py

```diff
@@ -133,19 +142,41 @@ def push(self):
                                                      self.args.debug)
 
         else:
-            self.d.login(self.args.username, self.args.password)
-            for line in self.d.push(self.image, stream=True):
-                bar = json.loads(line)
-                status = bar['status']
-                if prevstatus != status:
-                    util.write_out(status, "")
-                if 'id' not in bar:
-                    continue
-                if status == "Uploading":
-                    util.write_out(bar['progress'] + " ")
-                elif status == "Push complete":
-                    pass
-                elif status.startswith("Pushing"):
-                    util.write_out("Pushing: " + bar['id'])
-
-                prevstatus = status
+            reg, _, tag = util.decompose(self.image)
+            if not tag:
+                raise ValueError("The image being pushed must have a tag")
+
+            if self.args.password and self.args.username:
+                print(self.d.login(username=self.args.username, password=self.args.password, registry=reg))
+
+            sign_local = True
+            local_image = "docker-daemon:{}".format(self.image)
+            if self.args.reg_type == "atomic":
+                remote_image = "atomic:{}".format(self.image)
+                sign_local = False
+            else:
+                remote_image = "docker://{}".format(self.image)
+
+            sign = True if self.args.sign_by else False
+            if sign:
+                util.write_out("\nSigning with '{}'\n".format(self.args.sign_by))
+
+            insecure = True if util.is_insecure_registry(self.d.info()['RegistryConfig'], util.remove_port(reg)) else False
+
+            # We must push the file to the registry first prior to performing a
+            # local signature because the manifest file must be on the registry
+            return_code = util.skopeo_copy(local_image, remote_image, debug=self.args.debug,
+                                           sign_by=self.args.sign_by if not sign_local else None, insecure=insecure)
+
+            if return_code != 0:
+                raise ValueError("Pushing {} failed.".format(self.image))
+
+            util.write_out("Pushed: {}".format(self.image))
+
+            if sign_local and sign:
+                if self.args.debug:
+                    util.write_out("Need to create a local signature")
+                atomic_sign = Sign()
+                atomic_sign.sign(debug=self.args.debug, signer=self.args.sign_by,
```

Comment:
We could just do
atomic_sign.args.sign_by=self.args.sign_by
...

Rather then adding all of these parameters.